src/csync/csync_exclude.cpp

```diff
@@ -316,110 +307,265 @@ static CSYNC_EXCLUDE_TYPE _csync_excluded_common(c_strlist_t *excludes, const ch
         }
     }
 
-    if( ! excludes ) {
-        goto out;
-    }
+  out:
+    return match;
+}
 
-    if (check_leading_dirs) {
-        /* Build a list of path components to check. */
-        path_components = c_strlist_new(32);
-        char *path_split = strdup(path);
-        size_t len = strlen(path_split);
-        for (i = len; ; --i) {
-            // read backwards until a path separator is found
-            if (i != 0 && path_split[i-1] != '/') {
-                continue;
-            }
 
-            // check 'basename', i.e. for "/foo/bar/fi" we'd check 'fi', 'bar', 'foo'
-            if (path_split[i] != 0) {
-                c_strlist_add_grow(&path_components, path_split + i);
-            }
 
-            if (i == 0) {
-                break;
-            }
+/* Only for bnames (not paths) */
+static QString convertToBnameRegexpSyntax(QString exclude)
+{
+    QString s = "^" + QRegularExpression::escape(exclude).replace("\\*", ".*").replace("?", ".") + "$";
+    //qDebug() << "Converted pattern" << exclude << "to regex" << s;
+    return s;
+}
 
-            // check 'dirname', i.e. for "/foo/bar/fi" we'd check '/foo/bar', '/foo'
-            path_split[i-1] = '\0';
-            c_strlist_add_grow(&path_components, path_split);
-        }
-        SAFE_FREE(path_split);
-    }
 
-    /* Loop over all exclude patterns and evaluate the given path */
-    for (i = 0; match == CSYNC_NOT_EXCLUDED && i < excludes->count; i++) {
-        bool match_dirs_only = false;
-        char *pattern = excludes->vector[i];
+void csync_exclude_traversal_prepare(CSYNC *ctx)
+{
+    if (!ctx->excludes) {
+        qDebug() << "FURUZ";
+        //ctx->parsed_traversal_excludes.list_patterns_with_slashes = c_strlist_new(0);
+        return;
+    }
 
-        type = CSYNC_FILE_EXCLUDE_LIST;
-        if (!pattern[0]) { /* empty pattern */
+    QString _exclude_traversel_regexp_exclude;
+    QString _exclude_traversel_regexp_exclude_and_remove;
```

Comment:
There's a bug here: an empty regex will match anything. Check my patchset for how to start out with a match-nothing regex!